memtable/wbwi_memtable.cc

```diff
@@ -0,0 +1,147 @@
+    uint64_t tag = ExtractInternalKeyFooter(iter->key());
+    ValueType type;
+    SequenceNumber seq;
+    UnPackSequenceAndType(tag, &seq, &type);
+    // Unsupported operations.
+    assert(type != kTypeBlobIndex);
+    assert(type != kTypeWideColumnEntity);
+    assert(type != kTypeValuePreferredSeqno);
+    assert(type != kTypeDeletionWithTimestamp);
+    assert(type != kTypeMerge);
+    if (!callback || callback->IsVisible(seq)) {
+      if (*out_seq == kMaxSequenceNumber) {
+        *out_seq = std::max(seq, *max_covering_tombstone_seq);
+      }
+      if (*max_covering_tombstone_seq > seq) {
+        type = kTypeRangeDeletion;
+      }
+      switch (type) {
+        case kTypeValue: {
+          HandleTypeValue(lookup_user_key, iter->value(), iter->IsValuePinned(),
+                          do_merge, s->IsMergeInProgress(), merge_context,
+                          moptions_.merge_operator, clock_,
+                          moptions_.statistics, moptions_.info_log, s, value,
+                          columns, is_blob_index);
+          assert(seq <= read_seq);
+          return /*found_final_value=*/true;
+        }
+        case kTypeDeletion:
+        case kTypeSingleDeletion:
+        case kTypeRangeDeletion: {
+          HandleTypeDeletion(lookup_user_key, s->IsMergeInProgress(),
+                             merge_context, moptions_.merge_operator, clock_,
+                             moptions_.statistics, moptions_.info_log, s, value,
+                             columns);
+          assert(seq <= read_seq);
+          return /*found_final_value=*/true;
+        }
+        default: {
+          std::string msg("Unrecognized or unsupported value type: " +
+                          std::to_string(static_cast<int>(type)) + ". ");
+          msg.append("User key: " +
+                     ExtractUserKey(iter->key()).ToString(/*hex=*/true) + ". ");
+          msg.append("seq: " + std::to_string(seq) + ".");
+          *s = Status::Corruption(msg.c_str());
+          return /*found_final_value=*/true;
+        }
+      }
+    }
+    // Current key not visible or we read a merge key
+    assert(s->IsMergeInProgress() || (callback && !callback->IsVisible(seq)));
+    iter->Next();
+  }
+  if (!iter->status().ok() &&
+      (s->ok() || s->IsMergeInProgress() || s->IsNotFound())) {
+    *s = iter->status();
+    // stop further look up
+    return true;
+  }
+  return /*found_final_value=*/false;
+}
+
+void WBWIMemTable::MultiGet(const ReadOptions& read_options,
+                            MultiGetRange* range, ReadCallback* callback,
+                            bool immutable_memtable) {
+  (void)immutable_memtable;
+  // Should only be used as immutable memtable.
+  assert(immutable_memtable);
+  for (auto iter = range->begin(); iter != range->end(); ++iter) {
+    SequenceNumber dummy_seq;
+    bool found_final_value =
+        Get(*iter->lkey, iter->value ? iter->value->GetSelf() : nullptr,
```

Comment:
Maybe a TODO for future improvement is to re use the `InternalIterator` created from the memtable for MultiGet.

Reply:
good idea, I think we can do the same for MemTable too.